### PR TITLE
Refactored more of Entities and BlockEntities to use Vector3.

### PR DESCRIPTION
**Pull request by madmaxoft**
The BlockEntities use `Vector3i` for position storage.
Both BlockEntities and Entities use `Vector3` in constructor.

Notably, added new utility functions to `Vector3`:
`added(ofs)` returns a copy of the vector moved by the specified offset (3 axes)
`addedX(ofs)` returns a copy of the vector moved by ofs in the X axis
`addedXZ(ofsX, ofsZ)` returns a copy of the vector moved by the specified offsets in X and Z axes.
etc.
These are used when iterating over the neighbors of existing `Vector3` instances.

Despite the looks of it, this *is* a step towards 1.13 :)

### CODE REVIEW

File: src/Entities/Pickup.h [hunk 1]
```diff
@@ -18,14 +18,15 @@ class cPlayer;
...
 	// tolua_end
 
+	using super = cEntity;
+
+public:  // tolua_export
+
 	CLASS_PROTODEF(cPickup)
 
-	cPickup(double a_PosX, double a_PosY, double a_PosZ, const cItem & a_Item, bool IsPlayerCreated, float a_SpeedX = 0.f, float a_SpeedY = 0.f, float a_SpeedZ = 0.f, int a_LifetimeTicks = 6000, bool a_CanCombine = true);
+	cPickup(Vector3d a_Pos, const cItem & a_Item, bool IsPlayerCreated, Vector3f a_Speed = Vector3d(), int a_LifetimeTicks = 6000, bool a_CanCombine = true);
```
Comment (peterbell10) [inline src/Entities/Pickup.h]:
```suggestion
	cPickup(Vector3d a_Pos, const cItem & a_Item, bool IsPlayerCreated, Vector3f a_Speed = Vector3f(), int a_LifetimeTicks = 6000, bool a_CanCombine = true);
```

File: src/Items/ItemFishingRod.h [hunk 2]
```diff
@@ -245,7 +245,7 @@ class cItemFishingRodHandler :
 		}
 		else
 		{
-			auto Floater = cpp14::make_unique<cFloater>(a_Player->GetPosX(), a_Player->GetStance(), a_Player->GetPosZ(), a_Player->GetLookVector() * 15, a_Player->GetUniqueID(), (Random.RandInt(100, 900) - static_cast<int>(a_Player->GetEquippedItem().m_Enchantments.GetLevel(cEnchantments::enchLure) * 100)));
+			auto Floater = cpp14::make_unique<cFloater>(a_Player->GetPosition().addedY(1.5), a_Player->GetLookVector() * 15, a_Player->GetUniqueID(), (Random.RandInt(100, 900) - static_cast<int>(a_Player->GetEquippedItem().m_Enchantments.GetLevel(cEnchantments::enchLure) * 100)));
```
Comment (peterbell10) [inline src/Items/ItemFishingRod.h]: Magic number

File: src/BlockEntities/HopperEntity.cpp [hunk 3]
```diff
@@ -537,43 +532,37 @@ bool cHopperEntity::MoveItemsToChest(cChunk & a_Chunk, int a_BlockX, int a_Block
...
+		auto Block = neighbor->GetBlock(otherHalfRelCoord);
 		if (Block != ConnectedChest->GetBlockType())
 		{
 			// Not the same kind of chest
 			continue;
 		}
 
-		cChestEntity * Chest = static_cast<cChestEntity *>(Neighbor->GetBlockEntity(a_BlockX + Coords[i].x, a_BlockY, a_BlockZ + Coords[i].z));
-		if (Chest == nullptr)
+		auto chest = static_cast<cChestEntity *>(neighbor->GetBlockEntity(a_Coords.addedXZ(ofs.x, ofs.z)));
```
Comment (peterbell10) [inline src/BlockEntities/HopperEntity.cpp]:
```suggestion
		auto chest = static_cast<cChestEntity *>(neighbor->GetBlockEntity(a_Coords + ofs)));
```

File: src/BlockEntities/HopperEntity.cpp [hunk 4]
```diff
@@ -537,43 +532,37 @@ bool cHopperEntity::MoveItemsToChest(cChunk & a_Chunk, int a_BlockX, int a_Block
...
-	for (size_t i = 0; i < ARRAYCOUNT(Coords); i++)
-	{
-		int x = RelX + Coords[i].x;
-		int z = RelZ + Coords[i].z;
-		cChunk * Neighbor = a_Chunk.GetRelNeighborChunkAdjustCoords(x, z);
-		if (Neighbor == nullptr)
+	auto relCoord = a_Chunk.AbsoluteToRelative(a_Coords);
+	for (const auto & ofs: neighborOfs)
+	{
+		auto otherHalfRelCoord = relCoord.addedXZ(ofs.x, ofs.z);
```
Comment (peterbell10) [inline src/BlockEntities/HopperEntity.cpp]:
```suggestion
		auto otherHalfRelCoord = relCoord + ofs;
```

File: src/BlockEntities/HopperEntity.cpp [hunk 5]
```diff
@@ -537,43 +532,37 @@ bool cHopperEntity::MoveItemsToChest(cChunk & a_Chunk, int a_BlockX, int a_Block
...
 			continue;
 		}
 
-		cChestEntity * Chest = static_cast<cChestEntity *>(Neighbor->GetBlockEntity(a_BlockX + Coords[i].x, a_BlockY, a_BlockZ + Coords[i].z));
-		if (Chest == nullptr)
+		auto chest = static_cast<cChestEntity *>(neighbor->GetBlockEntity(a_Coords.addedXZ(ofs.x, ofs.z)));
+		if (chest == nullptr)
 		{
-			FLOGWARNING("{0}: A chest entity was not found where expected, at {1} ({2}, {3}})", __FUNCTION__, Vector3i{a_BlockX + Coords[i].x, a_BlockY, a_BlockZ + Coords[i].z}, x, z);
+			FLOGWARNING("{0}: A chest entity was not found where expected, at {1} ({2}, {3}})", __FUNCTION__, a_Coords.addedXZ(ofs.x, ofs.z), ofs.x, ofs.z);
```
Comment (peterbell10) [inline src/BlockEntities/HopperEntity.cpp]:
```suggestion
			FLOGWARNING("{0}: A chest entity was not found where expected, at {1} ({2}, {3}})", __FUNCTION__, a_Coords + ofs, ofs.x, ofs.z);
```

File: src/Vector3.h [hunk 6]
```diff
@@ -289,6 +289,36 @@ class Vector3
 		);
 	}
 
+	/** Returns a copy of this vector moved by the specified offset on all axes. */
+	inline Vector3<T> added(Vector3<T> a_Offset) const
```
Comment (peterbell10) [inline src/Vector3.h]: I think `added{X,Y,Z}` are a nice convenience but I can't see why you would use `added` or `addedXZ` over just `operator+`.

Comment (madmaxoft) [reply src/Vector3.h]: True, I thought I had a use case, but can't remember it now.

For the `XZ` version, we do have lots of code that scans only horizontal neighbors, not vertical, so they don't need to add the Y coord (and usually they don't even have one).

I'll get rid of the full `added`.

File: src/Items/ItemFishingRod.h (hunk 2 above)
Comment (madmaxoft) [reply src/Items/ItemFishingRod.h]: I knew `GetEyePosition()` was there, for some reason MSVC wouldn't add it to the suggestion list. Fixing...

**Pull request merged**